### ISSUE TITLE
Incorporating literature values for thermophysical properties of materials

### ISSUE THREAD

**Issue by AvonReppert** [dschick/udkm1DsimML#2]
udkm1Dsim provides the possibility to use temperature dependent parameters for thermophysical properties (e.g. heat capacity, heat conductivity and linear thermal expansion coefficients) in the n-temperature model when they exist as analytical functions. 

However very often one would like to use (experimentally determined) literature values for which often no explicit formula is available. If that is the case one needs to restort to clumsy workarounds such as writing functions that look up the values in tables, which dramatically increases the simulation speed. 

One example for a code snipped for such a workaround is attached in the .txt file
[workaroundCpFePt.txt](https://github.com/dschick/udkm1Dsim/files/1743388/workaroundCpFePt.txt)
It uses a table of values in which it looks for the heat capacity and integral of the heat capacity at the requested temperature. 

It would be nice to have an interface to where one would simply input the existing tabulated values which then does the interpolation/extrapolation.

Another possibilty is an automated, most likely section-wise parametrization of the properties of interest using polynominals. This would be more difficult I guess but possibly faster than "table-lookup" - functions.

**Comment by dschick** [comment]
Generally, I would also push this to collecting such data sets in the udkm1Dsim as some kind of database, similar to the X-ray parameters.

One could calculate e.g. the heat capacity and its integral in the beginning of the heat simulation on a very fine temperature grid using a interpolation method from experimental data and store that in the RAM. In the actutal simulation one would then use the finderb-method to get the best fitting data point.

So here is an idea: 

- We add a folder for heat capacity, thermal conductivity and thermal expansion coefficients to the parameters folders.
- there one can put tabulated files for the temperture dependend properties + a reference!
- the file name must be equal to the Name of the unitCell - not its ID
- the unitCell class will then a lookup in that folder for a file and use it if present or use the currently implemented way of property definition
- for the simulation it would be best, if one could estimate the highest and lowest temperture (latter should be okay) which is necessary +-50K or so in order to minimize the needed memory
- then in the simulation the given functions from the text files have to be interpolated onto a fine temperature grid, let's say 0.1K and numerically integrated
- the results could be stored into ram or maybe also as additional colums into the original text files or simply as additional text files with some suffix in the filename.
- if we store the interpolated data into textfiles we simply need to read them in the beginning and store them into ram
- in the simulation one can then use finderb which might be as fast as evaluating a complex analytical function.

I have some concerns with that:
* if we do not know an upper limit of the temperatur range we have to interpolate to quite high values which will result in relatively large files and arrays stored to ram making things slow again
* what if the experimental data is not defined on the necessary temperature range? we will need a warning/error message telling the user that problem. 
* how do we deal with the N-temperature model. do we want to have this feature also availabe for that?

**Comment by AvonReppert** [comment]
Yes such a database would be a very nice and useful feature that would greatly enhance the usability of the toolbox. 

However it would be a step by step process to build this up. I have some selected data for Nickel, Dysproisum/Holmium, Gold and Magnetite as a start but this would be a very slow process of building up the database. 

This database could also be enhanced with the complex refractive indices and their wavelength dependence ... 
Is an automated readout of [https://refractiveindex.info/](https://refractiveindex.info/) possible? 

In principal the heat capacity, the thermal expansion can be seperated into phonon, spin and electron contributions. See for example: [Ni: Meschter, Peter J., et al. "Physical contributions to the heat capacity of nickel." Journal of Physics and Chemistry of Solids 42.9 (1981): 861-871.](https://www.sciencedirect.com/science/article/pii/0022369781901748) and [Ni,Cu, Au: Nix, F. C., and D. MacNair.  "The thermal expansion of pure metals: copper, gold, aluminum, nickel, and iron." Physical Review 60.8 (1941): 597](https://journals.aps.org/pr/abstract/10.1103/PhysRev.60.597)

So in principal this could be done for an N-Temperature models for selected materials where this is possible. This would include a larger effort. Maybe in the material parameter folders on would have one folder for a 1 Temperature model and 3 other folder for electron, spin, and phonon systems seperately.
